cmd/kwild/server/migration.go

```diff
@@ -0,0 +1,183 @@
+		return m.kwildCfg, m.genesisCfg, nil
+	}
+
+	// poll for the genesis state
+	if err = m.pollForGenesisState(ctx); err != nil {
+		return nil, nil, err
+	}
+
+	return m.kwildCfg, m.genesisCfg, nil
+}
+
+// pollForGenesisState polls for the genesis state from the old chain
+// and downloads the genesis state to the snapshot file abnd updates the genesis config
+// and the kwild config with the configuration required for migration
+func (m *migrationClient) pollForGenesisState(ctx context.Context) (err error) {
+	// Poll for the genesis state from the old chain
+	m.logger.Info("Requesting genesis state from the old chain", log.String("listen_address", m.listenAddress))
+	for {
+		if err = m.downloadGenesisState(ctx); err == nil {
+			return nil
+		}
+
+		// retry after defaultPollFrequency
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(defaultPollFrequency):
+			m.logger.Info("Genesis state not available", log.Error(err), log.Duration("retry after (sec)", defaultPollFrequency))
+		}
+	}
+}
+
+func (m *migrationClient) downloadGenesisState(ctx context.Context) error {
+
+	metadata, err := m.clt.GenesisState(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Check if the genesis state is ready
+	if metadata.MigrationState.Status == types.NoActiveMigration || metadata.MigrationState.Status == types.MigrationNotStarted {
+		return fmt.Errorf("status %s", metadata.MigrationState.Status.String())
+	}
+
+	// Genesis state should ready
+	if metadata.SnapshotMetadata == nil || metadata.GenesisConfig == nil {
+		return errors.New("genesis state not available")
+	}
+
+	var genCfg chain.GenesisConfig
+	if err := json.Unmarshal(metadata.GenesisConfig, &genCfg); err != nil {
+		return fmt.Errorf("failed to unmarshal genesis config: %w", err)
+	}
+
+	// Save the genesis state
+	var snapshotMetadata statesync.Snapshot
+	if err := json.Unmarshal(metadata.SnapshotMetadata, &snapshotMetadata); err != nil {
+		return fmt.Errorf("failed to unmarshal snapshot metadata: %w", err)
+	}
+
+	m.logger.Info("Genesis state available for download")
+
+	// create snapshot file
+	genesisStateFile, err := os.Create(m.snapshotFileName)
+	if err != nil {
+		return fmt.Errorf("failed to create genesis snapshot file: %w", err)
+	}
+
+	// retrieve all the snapshot chunks
+	for i := uint32(0); i < snapshotMetadata.ChunkCount; i++ {
+		chunk, err := m.clt.GenesisSnapshotChunk(ctx, snapshotMetadata.Height, i)
+		if err != nil {
+			return fmt.Errorf("failed to download genesis snapshot chunk: %d  error: %w", i, err)
+		}
+		n, err := genesisStateFile.Write(chunk)
+		if err != nil {
+			return fmt.Errorf("failed to write genesis snapshot chunk: %d  error: %w", i, err)
+		}
+		if n != len(chunk) {
+			return fmt.Errorf("incomplete write to genesis snapshot chunk. expected: %d, written: %d", len(chunk), n)
+		}
+	}
+
+	// Update the genesis config
+	m.genesisCfg.DataAppHash = genCfg.DataAppHash
```

Comment:
Don't we need to compute this from the downloaded state file?

Reply:
the genesis file downloaded already has this hash computed, avoiding recomputation here.